R/InitErgmTerm.R

```diff
@@ -182,30 +182,45 @@ InitErgmTerm.absdiff <- function(nw, arglist, ..., version=packageVersion("ergm"
 
 
 ################################################################################
-InitErgmTerm.absdiffcat <- function(nw, arglist, ...) {
-  ### Check the network and arguments to make sure they are appropriate.
-  a <- check.ErgmTerm(nw, arglist, directed=NULL, bipartite=NULL,
-                      varnames = c("attrname","base"),
-                      vartypes = c("character","numeric"),
-                      defaultvalues = list(NULL,NULL),
-                      required = c(TRUE,FALSE))
+InitErgmTerm.absdiffcat <- function(nw, arglist, ..., version=packageVersion("ergm")) {
+  if(version <= as.package_version("3.9.4")){
+    ### Check the network and arguments to make sure they are appropriate.
+    a <- check.ErgmTerm(nw, arglist, directed=NULL, bipartite=NULL,
+                        varnames = c("attrname","base"),
+                        vartypes = c("character","numeric"),
+                        defaultvalues = list(NULL,NULL),
+                        required = c(TRUE,FALSE))
+	attrarg <- a$attrname
+  }else{
+    a <- check.ErgmTerm(nw, arglist, directed=NULL, bipartite=NULL,
+                        varnames = c("attr","levels"),
+                        vartypes = c(ERGM_VATTR_SPEC,ERGM_LEVELS_SPEC),
+                        defaultvalues = list(NULL,NULL),
+                        required = c(TRUE,FALSE))
+    attrarg <- a$attr
+  }
   ### Process the arguments
-  nodecov <- get.node.attr(nw, a$attrname)
+  nodecov <- ergm_get_vattr(attrarg, nw, accept = "numeric")
+  attrname <- attr(nodecov, "name")
+
   u <- sort(unique(as.vector(abs(outer(nodecov,nodecov,"-")))),na.last=NA)
   u <- u[u>0]
   NAsubstitute <- 2*(1+max(abs(c(nodecov,u)),na.rm=TRUE)) # Arbitrary unused (and nonzero) value
   napositions <- is.na(nodecov)
   nodecov[napositions] <- NAsubstitute
   if(any(napositions)){u<-c(u,NA)}
+  
+  u <- ergm_attr_levels(a$levels, u, nw, levels = u)
+  
   if(any(NVL(a$base,0)!=0)) u <- u[-(a$base)]
```

Comment:
Can you check that `levels` and `base` are handled correctly in both the default mode and in the 3.9.4 mode?

Reply:
Appears fine to me.  If a$levels is NULL (as it would be in the 3.9.4 version), then the ergm_attr_levels call just results in u.  If a$base is NULL (as it would be in the new default), then any(NVL(a$base,0)!=0) is FALSE, so u <- u[-(a$base)] is never executed.

Perhaps it's debatable whether u or nodecov should be passed to ergm_attr_levels as the "attr" argument?  Since we're selecting levels from u I passed u as the "attr" (and the "levels") but we could pass nodecov as "attr" and u as "levels".


results using the code in the pull request:

> require(ergm)
> set.seed(234)
> nwu = network(200, directed = F, density = 0.1)
> 
> num = sample(1:5, 200, replace = T)
> nwu %v% "num" <- num
> 
> # the sorted vector of unique values u will be 1 2 3 4 in this case
> 
> # keep everything
> summary(nwu ~ absdiffcat("num"))
absdiff.num.1 absdiff.num.2 absdiff.num.3 absdiff.num.4 
          614           489           329           185 
> 
> # keep 3 and 4
> summary(nwu ~ absdiffcat("num", levels = 3:4))
absdiff.num.3 absdiff.num.4 
          329           185 
> 
> # keep 1 and 2
> summary(nwu ~ absdiffcat("num", levels = 1:2))
absdiff.num.1 absdiff.num.2 
          614           489 
> 
> # also keep 3 and 4 since u[i] = i here
> summary(nwu ~ absdiffcat("num", levels = I(3:4)))
absdiff.num.3 absdiff.num.4 
          329           185 
> 
> 
> options(ergm.term=list(version=as.package_version("3.9.4")))
> 
> 
> # keep everything
> summary(nwu ~ absdiffcat("num"))
absdiff.num.1 absdiff.num.2 absdiff.num.3 absdiff.num.4 
          614           489           329           185 
> 
> # base 1 and 2, keeping 3 and 4
> summary(nwu ~ absdiffcat("num", base = 1:2))
absdiff.num.3 absdiff.num.4 
          329           185 
> 
> # base 3 and 4, keeping 1 and 2
> summary(nwu ~ absdiffcat("num", base = 3:4))
absdiff.num.1 absdiff.num.2 
          614           489 
		  
		  


CRAN results:

		  
> set.seed(234)
> nwu = network(200, directed = F, density = 0.1)
> 
> num = sample(1:5, 200, replace = T)
> nwu %v% "num" <- num
> 
> # the sorted vector of unique values u will be 1 2 3 4 in this case
> 
> # keep everything
> summary(nwu ~ absdiffcat("num"))
absdiff.num.1 absdiff.num.2 absdiff.num.3 absdiff.num.4 
          614           489           329           185 
> 
> # base 1 and 2, keeping 3 and 4
> summary(nwu ~ absdiffcat("num", base = 1:2))
absdiff.num.3 absdiff.num.4 
          329           185 
> 
> # base 3 and 4, keeping 1 and 2
> summary(nwu ~ absdiffcat("num", base = 3:4))
absdiff.num.1 absdiff.num.2 
          614           489

Reply:
OK, that makes sense. By the way, you can use triple-backticks (\`\`\` CODE \`\`\`) to typeset complex code.